src/library/Cardano/Fee.hs

```diff
@@ -331,26 +347,84 @@ coverRemainingFee (Fee fee) = go [] where
 -- >>> reduceChangeOutputs (DustThreshold 0) (Fee 15) (Coin <$> [1, 2, 4, 8])
 -- []
 --
-reduceChangeOutputs :: DustThreshold -> Fee -> [Coin] -> [Coin]
-reduceChangeOutputs threshold (Fee totalFee) changeOutputs
-    | totalFee >= totalChange =
-        []
-    | otherwise =
-        case positiveChangeOutputs of
-            x : xs -> x :| xs
-                & distributeFee (Fee totalFee)
-                & fmap payFee
-                & coalesceDust threshold
-            [] -> []
-  where
-    payFee :: (Fee, Coin) -> Coin
-    payFee (Fee f, c) =
-        fromMaybe C.zero (c `C.sub` f)
-
-    positiveChangeOutputs :: [Coin]
-    positiveChangeOutputs = filter (> C.zero) changeOutputs
-
-    totalChange = F.fold changeOutputs
+reduceChangeOutputs
+    :: FeeOptions i o
+    -> CoinSelection i o
+    -> Either (FeeAdjustmentError i o) (CoinSelection i o, Fee)
+reduceChangeOutputs opts s = do
+    -- The original requested fee amount
+    let Fee φ_original = estimateFee (feeEstimator opts) s
+    -- The initial amount left for fee (i.e. inputs - outputs)
+    let mδ_original = sumInputs s `C.sub` (sumOutputs s `C.add` sumChange s)
+    case mδ_original of
+        -- selection is now balanced, nothing to do.
+        Just δ_original | φ_original == δ_original -> do
+            pure (s, Fee C.zero)
+
+        -- some fee left to pay, but we've depleted all change outputs
+        Just δ_original | φ_original > δ_original && null (change s) -> do
+            let Just remainder = φ_original `C.sub` δ_original -- cf above guard
```

Comment:
Similar to: https://github.com/input-output-hk/cardano-coin-selection/pull/62#discussion_r414346686

We could use `C.distance` here.